# RFP: Betting on the Bachelor

## Project Overview
You are invited to submit a proposal that answers the following question:

### Who will win season 29 of the Bachelor?

*All proposals must be submitted by **1/15/25 at 11:59 PM**.*

## Required Proposal Components

### 1. Data Description
In the code cell below, read in the data you plan on using to train and test your model. Call `info()` once you have read the data into a dataframe. Consider using some or all of the following sources:
- [Scrape Fandom Wikis](https://bachelor-nation.fandom.com/wiki/The_Bachelor) or [the official Bachelor website]('https://bachelornation.com/shows/the-bachelor/')
- [Ask ChatGPT to genereate it](https://chatgpt.com/)
- [Read in csv files like this](https://www.kaggle.com/datasets/brianbgonz/the-bachelor-contestants?select=contestants.csv)

*Note, a level 5 dataset contains at least 1000 rows of non-null data. A level 4 contains at least 500 rows of non-null data.*

In [47]:
import pandas as pd
df = pd.read_csv('bachelor-contestants.csv')
df = pd.read_csv('contestants_level4.csv')

display(df)

Unnamed: 0            Name   Age                   Hometown  ElimWeek  \
0             0    Amanda Marsh  23.0            Chanute, Kansas      11.0   
1             1     Trista Rehn  29.0             Miami, Florida       6.0   
2             2  Shannon Oliver  24.0              Dallas, Texas       5.0   
3             3             Kim  24.0             Tempe, Arizona       4.0   
4             4    Cathy Grimes  22.0       Terra Haute, Indiana       3.0   
..          ...             ...   ...                        ...       ...   
495          20        Kristina  27.0         Brooklyn, New York       1.0   
496          21        Kristine  23.0  Fairfax Station, Virginia       1.0   
497          22         Kyshawn  30.0       Nashville, Tennessee       1.0   
498          23         Siomara  25.0          Chicago, Illinois       1.0   
499          24         Valerie  28.0          Wayne, New Jersey       1.0   

     Season  Region  
0         1       4  
1         1       2  
2         1       2  
3         1       4  
4         1       4  
..      ...     ...  
495       7       3  
496       7       2  
497       7       2  
498       7       4  
499       7       3  

[500 rows x 7 columns]

### 2. Training Your Model
In the cell seen below, write the code you need to train a linear regression model. Make sure you display the equation of the plane that best fits your chosen data at the end of your program. 

*Note, level 5 work trains a model using only the standard Python library and Pandas. A level 5 model is trained with at least two features, where one of the features begins as a categorical value (e.g. occupation, hometown, etc.). A level 4 uses external libraries like scikit or numpy.*

In [48]:
# Train model here.
# Don't forget to display the equation of the plane that best fits your data!
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
df.info()
display(df)
X = df[["Age", "Region"]]
y = df["ElimWeek"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
model = LinearRegression()
model.fit(X_train, y_train)
slope = model.coef_[0]
slope2 = model.coef_[1]
intercept = model.intercept_
print(slope, slope2, intercept) # y = -0.09x + 5.5, where x is age and y is elim week

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  500 non-null    int64  
 1   Name        500 non-null    object 
 2   Age         500 non-null    float64
 3   Hometown    500 non-null    object 
 4   ElimWeek    500 non-null    float64
 5   Season      500 non-null    int64  
 6   Region      500 non-null    int64  
dtypes: float64(2), int64(3), object(2)
memory usage: 27.5+ KB


Unnamed: 0            Name   Age                   Hometown  ElimWeek  \
0             0    Amanda Marsh  23.0            Chanute, Kansas      11.0   
1             1     Trista Rehn  29.0             Miami, Florida       6.0   
2             2  Shannon Oliver  24.0              Dallas, Texas       5.0   
3             3             Kim  24.0             Tempe, Arizona       4.0   
4             4    Cathy Grimes  22.0       Terra Haute, Indiana       3.0   
..          ...             ...   ...                        ...       ...   
495          20        Kristina  27.0         Brooklyn, New York       1.0   
496          21        Kristine  23.0  Fairfax Station, Virginia       1.0   
497          22         Kyshawn  30.0       Nashville, Tennessee       1.0   
498          23         Siomara  25.0          Chicago, Illinois       1.0   
499          24         Valerie  28.0          Wayne, New Jersey       1.0   

     Season  Region  
0         1       4  
1         1       2  
2         1       2  
3         1       4  
4         1       4  
..      ...     ...  
495       7       3  
496       7       2  
497       7       2  
498       7       4  
499       7       3  

[500 rows x 7 columns]

-0.06578016038585983 -0.22266259804401312 5.769421903003378


### 3. Testing Your Model
In the cell seen below, write the code you need to test your linear regression model. 

*Note, a model is considered a level 5 if it achieves at least 60% prediction accuracy or achieves an RMSE of 2 weeks or less.*

In [57]:
squared_error_total = 0

# Iterate over the DataFrame rows
for i in range(len(df)):
    # Projected value (calculated using the formula)
    projected_value = (
        float(df.iloc[i, 2]) * -0.06578016038585983 +  # Third column
        float(df.iloc[i, 6]) * -0.22266259804401312 +  # Last column (7th column)
        5.769421903003378                              # Constant
    )
    
    # Actual value from the fifth column
    actual_value = float(df.iloc[i, 4])
    
    # Calculate squared error for this row
    squared_error = (projected_value - actual_value) ** 2
    
    # Add to total squared error
    squared_error_total += squared_error

# Calculate mean squared error
mse = squared_error_total / len(df) if len(df) > 0 else None

# Print the MSE
if mse is not None:
    print(f"Mean Squared Error: {mse}")
else:
    print("DataFrame is empty")

Mean Squared Error: 7.365757559463378


### 4. Final Answer

In the first cell seen below, state the name of your predicted winner. 
In the second cell seen below, justify your prediction using an evaluation technique like RMSE or percent accuracy.

#### State the name of your predicted winner here.
Don't know and am not going to answer

#### Justify your prediction here.
MEAN SQUARED ERROR OF 7.37 I CAN'T PREDICT ANYTHING